longitudinal_controller/src/longitudinal_controller.py

```diff
@@ -0,0 +1,227 @@
+                    throttle_diff, p, i, d = self.pid_controller.update(
+                        error, self.anti_windup_enabled, dt
+                    )
+                    rospy.logwarn("Throttle diff: " + str(throttle_diff))
+
+                    # Add the controller's output value to the idle position
+                    throttle_value = self.throttle_idle_autonomous_pwm + throttle_diff
+
+                    #  --- Anti windup strategy ---
+
+                    # Check if throttle output is already saturating the actuator
+                    if (
+                        throttle_value < self.throttle_min_autonomous_pwm
+                        or throttle_value > self.throttle_max_autonomous_pwm
+                    ):
+                        throttle_saturating = True
+
+                        if throttle_value < self.throttle_min_autonomous_pwm:
+                            throttle_value = self.throttle_min_autonomous_pwm
+
+                        elif throttle_value > self.throttle_max_autonomous_pwm:
+                            throttle_value = self.throttle_max_autonomous_pwm
+
+                    else:
+                        throttle_saturating = False
+
+                    # Check if controller is trying to saturate the actuator even more
+                    if np.sign(error) == np.sign(throttle_value):
+                        controller_saturating = True
+                    else:
+                        controller_saturating = False
+
+                    # If both above are true, then stop integrating.
+                    if throttle_saturating and controller_saturating:
+                        self.anti_windup_enabled = True
+                    else:
+                        self.anti_windup_enabled = False
+
+                    rospy.logwarn("Anti windup: " + str(self.anti_windup_enabled))
+
+                    #  Publish result
+                    self.publish_throttle_cmd(throttle_value)
+                    self.publish_debug_pid(p, i, d)
+                    self.previous_t = time.time()
+
+                else:
+                    # We are in PID mode but the controller is disabled because the engage button
+                    # is not pressed (or value not received since for too long)
+                    rospy.logwarn("Controller OFF")
+
+                    # Publish the throttle IDLE value
+                    self.publish_throttle_cmd(self.throttle_idle_autonomous_pwm)
+                    # Reset the integral component
+                    self.pid_controller.reset()
+                    # Reset the previous time to avoid having a large integral value when we engage again
+                    self.previous_t = time.time()
+
+            elif (
+                self.longitudinal_control_mode
+                == LongitudinalControlMode.ConstantPwmOutput.value
+            ):
+
+                self.publish_throttle_cmd(self.constant_pwm_output)
+
```

Comment:
Shall we set `self.previous_t = time.time()` here too? Otherwise the first `dt` when going from ConstantPwmOutput to PID (I we ever do that with dynamic reconfigure for instance) will be very large.

Reply:
Indeed. Putting it right before the rate.sleep() because we want to update it in every case anyway.